File: desktop/main/connection-state.ts
/**
 * Pure connection state machine for the attach-only desktop pet (U6).
 *
 * No child_process, PID, signal, or service ownership — only probe/attach UX.
 */

export const DESKTOP_DEFAULT_PORT = 62666;
export const DESKTOP_DEFAULT_HOST = "127.0.0.1";
export const DESKTOP_START_COMMAND = "spi --no-open";

export type DesktopConnectionStatus =
  | "probing"
  | "connected"
  | "reconnecting"
  | "service-not-running"
  | "incompatible";

/** Stable diagnostic codes for incompatible / failed attach (never raw HTML bodies). */
export type DesktopConnectionReasonCode =
  | "connection_refused"
  | "network_error"
  | "health_invalid"
  | "health_http_error"
  | "protocol_http_error"
  | "protocol_invalid"
  | "protocol_mismatch"
  | "product_mismatch"
  | "server_mode"
  | "auth_required"
  | "auth_invalid"
  | "token_rejected"
  | "instance_changed"
  | "stream_error"
  | "unknown";

export type DesktopConnectionState = {
  status: DesktopConnectionStatus;
  origin: string;
  port: number;
  /** Copyable only — never executed by the pet. */
  startCommand: string;
  instanceId: string | null;
  reasonCode: DesktopConnectionReasonCode | null;
  detail: string | null;
  attempt: number;
  /** When true, the next successful snapshot is a notification baseline (no toasts). */
  resetNotificationBaseline: boolean;
  /**
   * Additive protocol capability. Missing/unknown on old servers stays false so
   * observer attach still works while the quick-session entry stays hidden.
   */
  quickSessionAvailable: boolean;
  updatedAt: number;
};

export type DesktopConnectionEvent =
  | { type: "start_probe" }
  | { type: "retry" }
  | { type: "connection_refused" }
  | { type: "network_error"; detail?: string }
  | {
      type: "incompatible";
      reasonCode: Exclude<
        DesktopConnectionReasonCode,
        "connection_refused" | "network_error" | "instance_changed" | "stream_error"
      >;
      detail?: string;
    }
  | {
      type: "connected";
      instanceId: string;
      resetBaseline?: boolean;
      quickSessionAvailable?: boolean;
    }
  | { type: "stream_lost"; detail?: string }
  | { type: "instance_changed"; instanceId: string }
  | { type: "token_rejected" }
  | { type: "quit" };

export function buildDesktopOrigin(port: number = DESKTOP_DEFAULT_PORT): string {
  const safePort = Number.isFinite(port) && port > 0 && port <= 65535 ? Math.floor(port) : DESKTOP_DEFAULT_PORT;
  return `http://${DESKTOP_DEFAULT_HOST}:${safePort}`;
}

export function createInitialConnectionState(input?: {
  port?: number;
  now?: number;
}): DesktopConnectionState {
  const port =
    input?.port && Number.isFinite(input.port) && input.port > 0 && input.port <= 65535
      ? Math.floor(input.port)
      : DESKTOP_DEFAULT_PORT;
  return {
    status: "probing",
    origin: buildDesktopOrigin(port),
    port,
    startCommand: DESKTOP_START_COMMAND,
    instanceId: null,
    reasonCode: null,
    detail: null,
    attempt: 0,
    resetNotificationBaseline: true,
    quickSessionAvailable: false,
    updatedAt: input?.now ?? 0,
  };
}

function withMeta(
  state: DesktopConnectionState,
  patch: Partial<DesktopConnectionState>,
  now: number,
): DesktopConnectionState {
  return {
    ...state,
    ...patch,
    origin: buildDesktopOrigin(patch.port ?? state.port),
    startCommand: DESKTOP_START_COMMAND,
    updatedAt: now,
  };
}

/**
 * Pure reducer. Service-not-running is distinct from incompatible diagnostics.
 * Quit only clears local networking intent — never mutates a service.
 */
export function reduceConnectionState(
  state: DesktopConnectionState,
  event: DesktopConnectionEvent,
  now = Date.now(),
): DesktopConnectionState {
  switch (event.type) {
    case "start_probe":
      return withMeta(
        state,
        {
          status: "probing",
          reasonCode: null,
          detail: null,
          attempt: state.attempt + (state.status === "probing" ? 0 : 1),
        },
        now,
      );

    case "retry":
      return withMeta(
        state,
        {
          status: "probing",
          reasonCode: null,
          detail: null,
          attempt: state.attempt + 1,
          // Manual retry after disconnect keeps baseline only when we were not connected.
          resetNotificationBaseline:
            state.status === "service-not-running" ||
            state.status === "incompatible" ||
            state.status === "probing"
              ? true
              : state.resetNotificationBaseline,
        },
        now,
      );

    case "connection_refused":
      return withMeta(
        state,
        {
          status: "service-not-running",
          reasonCode: "connection_refused",
          detail: null,
          instanceId: null,
          quickSessionAvailable: false,
        },
        now,
      );

    case "network_error":
      // Soft network blip while connected → reconnecting; cold probe → service-not-running
      // only when the failure looks like refusal is handled by the client via connection_refused.
      if (state.status === "connected" || state.status === "reconnecting") {
        return withMeta(
          state,
          {
            status: "reconnecting",
            reasonCode: "network_error",
            detail: clampDetail(event.detail),
            attempt: state.attempt + 1,
          },
          now,
        );
      }
      return withMeta(
        state,
        {
          status: "service-not-running",
          reasonCode: "network_error",
          detail: clampDetail(event.detail),
          instanceId: null,
          quickSessionAvailable: false,
        },
        now,
      );

    case "incompatible":
      return withMeta(
        state,
        {
          status: "incompatible",
          reasonCode: event.reasonCode,
          detail: clampDetail(event.detail),
          instanceId: null,
          quickSessionAvailable: false,
        },
        now,
      );

    case "connected": {
      const instanceChanged =
        state.instanceId != null && state.instanceId !== event.instanceId;
      const resetBaseline =
        event.resetBaseline === true ||
        state.resetNotificationBaseline ||
        instanceChanged ||
        state.status !== "connected";
      return withMeta(
        state,
        {
          status: "connected",
          instanceId: event.instanceId,
          reasonCode: null,
          detail: null,
          attempt: 0,
          resetNotificationBaseline: resetBaseline,
          quickSessionAvailable: event.quickSessionAvailable === true,
        },
        now,
      );
    }

    case "stream_lost":
      if (state.status === "service-not-running" || state.status === "incompatible") {
        return state;
      }
      return withMeta(
        state,
        {
          status: "reconnecting",
          reasonCode: "stream_error",
          detail: clampDetail(event.detail),
          attempt: state.attempt + 1,
        },
        now,
      );

    case "instance_changed":
      return withMeta(
        state,
        {
          status: "reconnecting",
          instanceId: event.instanceId,
          reasonCode: "instance_changed",
          detail: null,
          resetNotificationBaseline: true,
          quickSessionAvailable: false,
          attempt: state.attempt + 1,
        },
        now,
      );

    case "token_rejected":
      return withMeta(
        state,
        {
          status: "reconnecting",
          reasonCode: "token_rejected",
          detail: null,
          resetNotificationBaseline: true,
          attempt: state.attempt + 1,
        },
        now,
      );

    case "quit":
      // Local teardown marker only — does not imply service stop.
      return withMeta(
        state,
        {
          status: "service-not-running",
          reasonCode: null,
          detail: "desktop_quit",
          instanceId: null,
          attempt: 0,
          resetNotificationBaseline: true,
          quickSessionAvailable: false,
        },
        now,
      );

    default: {
      const _exhaustive: never = event;
      return _exhaustive;
    }
  }
}

/** After the first connected snapshot is applied for notifications, clear the baseline flag. */
export function acknowledgeConnectionBaseline(
  state: DesktopConnectionState,
  now = Date.now(),
): DesktopConnectionState {
  if (!state.resetNotificationBaseline) return state;
  return withMeta(state, { resetNotificationBaseline: false }, now);
}

export function isServiceNotRunning(state: DesktopConnectionState): boolean {
  return state.status === "service-not-running";
}

export function canCopyStartCommand(state: DesktopConnectionState): boolean {
  return state.status === "service-not-running" || state.status === "incompatible";
}

function clampDetail(detail: string | undefined | null): string | null {
  if (!detail) return null;
  const trimmed = detail.trim().replace(/\s+/g, " ");
  if (!trimmed) return null;
  return trimmed.length > 160 ? `${trimmed.slice(0, 159)}…` : trimmed;
}

File: desktop-tauri/src-tauri/src/connection_state.rs
use serde::{Deserialize, Serialize};

pub const DESKTOP_DEFAULT_PORT: u16 = 62666;
pub const DESKTOP_DEFAULT_HOST: &str = "127.0.0.1";
pub const DESKTOP_START_COMMAND: &str = "spi --no-open";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DesktopConnectionStatus {
    Probing,
    Connected,
    Reconnecting,
    ServiceNotRunning,
    Incompatible,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DesktopConnectionReasonCode {
    ConnectionRefused,
    NetworkError,
    HealthInvalid,
    HealthHttpError,
    ProtocolHttpError,
    ProtocolInvalid,
    ProtocolMismatch,
    ProductMismatch,
    ServerMode,
    AuthRequired,
    AuthInvalid,
    TokenRejected,
    InstanceChanged,
    StreamError,
    TlsError,
    RemoteUnsupported,
    InsecureHttpRejected,
    Unknown,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopConnectionState {
    pub status: DesktopConnectionStatus,
    pub origin: String,
    pub port: u16,
    pub start_command: String,
    pub instance_id: Option<String>,
    pub reason_code: Option<DesktopConnectionReasonCode>,
    pub detail: Option<String>,
    pub attempt: u32,
    pub reset_notification_baseline: bool,
    pub quick_session_available: bool,
    pub updated_at: i64,
    #[serde(default, skip_serializing)]
    pub generation: u64,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ConnectionTarget {
    pub profile_id: String,
    pub origin: String,
    pub port: u16,
    pub allow_insecure_http: bool,
    pub generation: u64,
}

impl ConnectionTarget {
    pub fn is_loopback(&self) -> bool {
        crate::server_profiles::is_loopback_origin(&self.origin)
    }

    pub fn is_remote(&self) -> bool {
        !self.is_loopback()
    }

    pub fn http_allowed(&self) -> bool {
        self.origin.starts_with("https://") || self.is_loopback() || self.allow_insecure_http
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum DesktopConnectionEvent {
    StartProbe,
    Retry,
    ConnectionRefused,
    NetworkError {
        #[serde(default)]
        detail: Option<String>,
    },
    Incompatible {
        #[serde(rename = "reasonCode")]
        reason_code: DesktopConnectionReasonCode,
        #[serde(default)]
        detail: Option<String>,
    },
    Connected {
        #[serde(rename = "instanceId")]
        instance_id: String,
        #[serde(rename = "resetBaseline")]
        #[serde(default)]
        reset_baseline: Option<bool>,
        #[serde(rename = "quickSessionAvailable")]
        #[serde(default)]
        quick_session_available: Option<bool>,
    },
    StreamLost {
        #[serde(default)]
        detail: Option<String>,
    },
    InstanceChanged {
        #[serde(rename = "instanceId")]
        instance_id: String,
    },
    TokenRejected,
    Quit,
}

pub fn sanitize_port(port: u16) -> u16 {
    if port == 0 {
        DESKTOP_DEFAULT_PORT
    } else {
        port
    }
}

pub fn build_desktop_origin(port: u16) -> String {
    format!("http://{DESKTOP_DEFAULT_HOST}:{}", sanitize_port(port))
}

pub fn is_loopback_observer_url(url: &str, port: u16) -> bool {
    crate::server_profiles::is_origin_scoped_url(url, &build_desktop_origin(port))
}

pub fn is_target_scoped_url(url: &str, origin: &str) -> bool {
    crate::server_profiles::is_origin_scoped_url(url, origin)
}

pub fn create_initial_connection_state(port: u16, now: i64) -> DesktopConnectionState {
    create_initial_connection_state_for_target(
        &ConnectionTarget {
            profile_id: crate::server_profiles::LOCAL_PROFILE_ID.to_string(),
            origin: build_desktop_origin(port),
            port: sanitize_port(port),
            allow_insecure_http: false,
            generation: 0,
        },
        now,
    )
}

pub fn create_initial_connection_state_for_target(
    target: &ConnectionTarget,
    now: i64,
) -> DesktopConnectionState {
    let loopback = target.is_loopback();
    DesktopConnectionState {
        status: DesktopConnectionStatus::Probing,
        origin: target.origin.clone(),
        port: sanitize_port(target.port),
        start_command: if loopback {
            DESKTOP_START_COMMAND.to_string()
        } else {
            String::new()
        },
        instance_id: None,
        reason_code: None,
        detail: None,
        attempt: 0,
        reset_notification_baseline: true,
        quick_session_available: false,
        updated_at: now,
        generation: target.generation,
    }
}

fn with_meta(
    state: &DesktopConnectionState,
    patch: DesktopConnectionState,
    now: i64,
) -> DesktopConnectionState {
    DesktopConnectionState {
        origin: state.origin.clone(),
        port: state.port,
        start_command: if crate::server_profiles::is_loopback_origin(&state.origin) {
            DESKTOP_START_COMMAND.to_string()
        } else {
            String::new()
        },
        generation: state.generation,
        updated_at: now,
        ..patch
    }
}

pub fn reduce_connection_state(
    state: &DesktopConnectionState,
    event: &DesktopConnectionEvent,
    now: i64,
) -> DesktopConnectionState {
    match event {
        DesktopConnectionEvent::StartProbe => with_meta(
            state,
            DesktopConnectionState {
                status: DesktopConnectionStatus::Probing,
                reason_code: None,
                detail: None,
                attempt: if state.status == DesktopConnectionStatus::Probing {
                    state.attempt
                } else {
                    state.attempt.saturating_add(1)
                },
                ..state.clone()
            },
            now,
        ),
        DesktopConnectionEvent::Retry => with_meta(
            state,
            DesktopConnectionState {
                status: DesktopConnectionStatus::Probing,
                reason_code: None,
                detail: None,
                attempt: state.attempt.saturating_add(1),
                reset_notification_baseline: matches!(
                    state.status,
                    DesktopConnectionStatus::ServiceNotRunning
                        | DesktopConnectionStatus::Incompatible
                        | DesktopConnectionStatus::Probing
                ) || state.reset_notification_baseline,
                ..state.clone()
            },
            now,
        ),
        DesktopConnectionEvent::ConnectionRefused => with_meta(
            state,
            DesktopConnectionState {
                status: DesktopConnectionStatus::ServiceNotRunning,
                reason_code: Some(DesktopConnectionReasonCode::ConnectionRefused),
                detail: None,
                instance_id: None,
                quick_session_available: false,
                ..state.clone()
            },
            now,
        ),
        DesktopConnectionEvent::NetworkError { detail } => {
            if matches!(
                state.status,
                DesktopConnectionStatus::Connected | DesktopConnectionStatus::Reconnecting
            ) {
                with_meta(
                    state,
                    DesktopConnectionState {
                        status: DesktopConnectionStatus::Reconnecting,
                        reason_code: Some(DesktopConnectionReasonCode::NetworkError),
                        detail: clamp_detail(detail.as_deref()),
                        attempt: state.attempt.saturating_add(1),
                        ..state.clone()
                    },
                    now,
                )
            } else {
                with_meta(
                    state,
                    DesktopConnectionState {
                        status: DesktopConnectionStatus::ServiceNotRunning,
                        reason_code: Some(DesktopConnectionReasonCode::NetworkError),
                        detail: clamp_detail(detail.as_deref()),
                        instance_id: None,
                        quick_session_available: false,
                        ..state.clone()
                    },
                    now,
                )
            }
        }
        DesktopConnectionEvent::Incompatible { reason_code, detail } => with_meta(
            state,
            DesktopConnectionState {
                status: DesktopConnectionStatus::Incompatible,
                reason_code: Some(*reason_code),
                detail: clamp_detail(detail.as_deref()),
                instance_id: None,
                quick_session_available: false,
                ..state.clone()
            },
            now,
        ),
        DesktopConnectionEvent::Connected {
            instance_id,
            reset_baseline,
            quick_session_available,
        } => {
            let instance_changed = state
                .instance_id
                .as_ref()
                .is_some_and(|current| current != instance_id);
            let reset = reset_baseline.unwrap_or(false)
                || state.reset_notification_baseline
                || instance_changed
                || state.status != DesktopConnectionStatus::Connected;
            with_meta(
                state,
                DesktopConnectionState {
                    status: DesktopConnectionStatus::Connected,
                    instance_id: Some(instance_id.clone()),
                    reason_code: None,
                    detail: None,
                    attempt: 0,
                    reset_notification_baseline: reset,
                    quick_session_available: quick_session_available.unwrap_or(false),
                    ..state.clone()
                },
                now,
            )
        }
        DesktopConnectionEvent::StreamLost { detail } => {
            if matches!(
                state.status,
                DesktopConnectionStatus::ServiceNotRunning | DesktopConnectionStatus::Incompatible
            ) {
                return state.clone();
            }
            with_meta(
                state,
                DesktopConnectionState {
                    status: DesktopConnectionStatus::Reconnecting,
                    reason_code: Some(DesktopConnectionReasonCode::StreamError),
                    detail: clamp_detail(detail.as_deref()),
                    attempt: state.attempt.saturating_add(1),
                    ..state.clone()
                },
                now,
            )
        }
        DesktopConnectionEvent::InstanceChanged { instance_id } => with_meta(
            state,
            DesktopConnectionState {
                status: DesktopConnectionStatus::Reconnecting,
                instance_id: Some(instance_id.clone()),
                reason_code: Some(DesktopConnectionReasonCode::InstanceChanged),
                detail: None,
                reset_notification_baseline: true,
                quick_session_available: false,
                attempt: state.attempt.saturating_add(1),
                ..state.clone()
            },
            now,
        ),
        DesktopConnectionEvent::TokenRejected => with_meta(
            state,
            DesktopConnectionState {
                status: DesktopConnectionStatus::Reconnecting,
                reason_code: Some(DesktopConnectionReasonCode::TokenRejected),
                detail: None,
                reset_notification_baseline: true,
                attempt: state.attempt.saturating_add(1),
                ..state.clone()
            },
            now,
        ),
        DesktopConnectionEvent::Quit => with_meta(
            state,
            DesktopConnectionState {
                status: DesktopConnectionStatus::ServiceNotRunning,
                reason_code: None,
                detail: Some("desktop_quit".to_string()),
                instance_id: None,
                attempt: 0,
                reset_notification_baseline: true,
                quick_session_available: false,
                ..state.clone()
            },
            now,
        ),
    }
}

pub fn acknowledge_connection_baseline(
    state: &DesktopConnectionState,
    now: i64,
) -> DesktopConnectionState {
    if !state.reset_notification_baseline {
        return state.clone();
    }
    with_meta(
        state,
        DesktopConnectionState {
            reset_notification_baseline: false,
            ..state.clone()
        },
        now,
    )
}

pub fn can_copy_start_command(state: &DesktopConnectionState) -> bool {
    crate::server_profiles::is_loopback_origin(&state.origin)
        && matches!(
            state.status,
            DesktopConnectionStatus::ServiceNotRunning | DesktopConnectionStatus::Incompatible
        )
}

pub fn public_connection_state(state: &DesktopConnectionState) -> serde_json::Value {
    serde_json::json!({
        "status": state.status,
        "origin": state.origin,
        "port": state.port,
        "startCommand": state.start_command,
        "instanceId": state.instance_id,
        "reasonCode": state.reason_code,
        "detail": state.detail,
        "attempt": state.attempt,
        "resetNotificationBaseline": state.reset_notification_baseline,
        "quickSessionAvailable": state.quick_session_available,
    })
}

fn clamp_detail(detail: Option<&str>) -> Option<String> {
    let trimmed = detail?
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    if trimmed.is_empty() {
        return None;
    }
    if trimmed.chars().count() > 160 {
        let clipped: String = trimmed.chars().take(159).collect();
        Some(format!("{clipped}…"))
    } else {
        Some(trimmed)
    }
}
